// @aws-sdk/client-pinpoint/dist-cjs/commands/GetRecommenderConfigurationsCommand.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetRecommenderConfigurationsCommand = void 0;
const middleware_serde_1 = require("@aws-sdk/middleware-serde");
const smithy_client_1 = require("@aws-sdk/smithy-client");
const models_1_1 = require("../models/models_1");
const Aws_restJson1_1 = require("../protocols/Aws_restJson1");
class GetRecommenderConfigurationsCommand extends smithy_client_1.Command {
    constructor(input) {
        super();
        this.input = input;
    }
    resolveMiddleware(clientStack, configuration, options) {
        this.middlewareStack.use((0, middleware_serde_1.getSerdePlugin)(configuration, this.serialize, this.deserialize));
        const stack = clientStack.concat(this.middlewareStack);
        const { logger } = configuration;
        const clientName = "PinpointClient";
        const commandName = "GetRecommenderConfigurationsCommand";
        const handlerExecutionContext = {
            logger,
            clientName,
            commandName,
            inputFilterSensitiveLog: models_1_1.GetRecommenderConfigurationsRequestFilterSensitiveLog,
            outputFilterSensitiveLog: models_1_1.GetRecommenderConfigurationsResponseFilterSensitiveLog,
        };
        const { requestHandler } = configuration;
        return stack.resolve((request) => requestHandler.handle(request.request, options || {}), handlerExecutionContext);
    }
    serialize(input, context) {
        return (0, Aws_restJson1_1.serializeAws_restJson1GetRecommenderConfigurationsCommand)(input, context);
    }
    deserialize(output, context) {
        return (0, Aws_restJson1_1.deserializeAws_restJson1GetRecommenderConfigurationsCommand)(output, context);
    }
}
exports.GetRecommenderConfigurationsCommand = GetRecommenderConfigurationsCommand;
