// @aws-sdk/client-pinpoint/dist-es/commands/CreateVoiceTemplateCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { CreateVoiceTemplateRequestFilterSensitiveLog, CreateVoiceTemplateResponseFilterSensitiveLog, } from "../models/models_0";
import { deserializeAws_restJson1CreateVoiceTemplateCommand, serializeAws_restJson1CreateVoiceTemplateCommand, } from "../protocols/Aws_restJson1";
var CreateVoiceTemplateCommand = (function (_super) {
    __extends(CreateVoiceTemplateCommand, _super);
    function CreateVoiceTemplateCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    CreateVoiceTemplateCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "PinpointClient";
        var commandName = "CreateVoiceTemplateCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: CreateVoiceTemplateRequestFilterSensitiveLog,
            outputFilterSensitiveLog: CreateVoiceTemplateResponseFilterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    CreateVoiceTemplateCommand.prototype.serialize = function (input, context) {
        return serializeAws_restJson1CreateVoiceTemplateCommand(input, context);
    };
    CreateVoiceTemplateCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_restJson1CreateVoiceTemplateCommand(output, context);
    };
    return CreateVoiceTemplateCommand;
}($Command));
export { CreateVoiceTemplateCommand };
