// @aws-sdk/client-pinpoint/dist-es/commands/DeleteApnsVoipChannelCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { DeleteApnsVoipChannelRequestFilterSensitiveLog, DeleteApnsVoipChannelResponseFilterSensitiveLog, } from "../models/models_0";
import { deserializeAws_restJson1DeleteApnsVoipChannelCommand, serializeAws_restJson1DeleteApnsVoipChannelCommand, } from "../protocols/Aws_restJson1";
var DeleteApnsVoipChannelCommand = (function (_super) {
    __extends(DeleteApnsVoipChannelCommand, _super);
    function DeleteApnsVoipChannelCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    DeleteApnsVoipChannelCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "PinpointClient";
        var commandName = "DeleteApnsVoipChannelCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: DeleteApnsVoipChannelRequestFilterSensitiveLog,
            outputFilterSensitiveLog: DeleteApnsVoipChannelResponseFilterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    DeleteApnsVoipChannelCommand.prototype.serialize = function (input, context) {
        return serializeAws_restJson1DeleteApnsVoipChannelCommand(input, context);
    };
    DeleteApnsVoipChannelCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_restJson1DeleteApnsVoipChannelCommand(output, context);
    };
    return DeleteApnsVoipChannelCommand;
}($Command));
export { DeleteApnsVoipChannelCommand };
