// @aws-sdk/client-pinpoint/dist-es/commands/GetApnsVoipChannelCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { GetApnsVoipChannelRequestFilterSensitiveLog, GetApnsVoipChannelResponseFilterSensitiveLog, } from "../models/models_0";
import { deserializeAws_restJson1GetApnsVoipChannelCommand, serializeAws_restJson1GetApnsVoipChannelCommand, } from "../protocols/Aws_restJson1";
var GetApnsVoipChannelCommand = (function (_super) {
    __extends(GetApnsVoipChannelCommand, _super);
    function GetApnsVoipChannelCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    GetApnsVoipChannelCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "PinpointClient";
        var commandName = "GetApnsVoipChannelCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: GetApnsVoipChannelRequestFilterSensitiveLog,
            outputFilterSensitiveLog: GetApnsVoipChannelResponseFilterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    GetApnsVoipChannelCommand.prototype.serialize = function (input, context) {
        return serializeAws_restJson1GetApnsVoipChannelCommand(input, context);
    };
    GetApnsVoipChannelCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_restJson1GetApnsVoipChannelCommand(output, context);
    };
    return GetApnsVoipChannelCommand;
}($Command));
export { GetApnsVoipChannelCommand };
