// @aws-sdk/client-pinpoint/dist-es/commands/GetApplicationDateRangeKpiCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { GetApplicationDateRangeKpiRequestFilterSensitiveLog, GetApplicationDateRangeKpiResponseFilterSensitiveLog, } from "../models/models_0";
import { deserializeAws_restJson1GetApplicationDateRangeKpiCommand, serializeAws_restJson1GetApplicationDateRangeKpiCommand, } from "../protocols/Aws_restJson1";
var GetApplicationDateRangeKpiCommand = (function (_super) {
    __extends(GetApplicationDateRangeKpiCommand, _super);
    function GetApplicationDateRangeKpiCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    GetApplicationDateRangeKpiCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "PinpointClient";
        var commandName = "GetApplicationDateRangeKpiCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: GetApplicationDateRangeKpiRequestFilterSensitiveLog,
            outputFilterSensitiveLog: GetApplicationDateRangeKpiResponseFilterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    GetApplicationDateRangeKpiCommand.prototype.serialize = function (input, context) {
        return serializeAws_restJson1GetApplicationDateRangeKpiCommand(input, context);
    };
    GetApplicationDateRangeKpiCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_restJson1GetApplicationDateRangeKpiCommand(output, context);
    };
    return GetApplicationDateRangeKpiCommand;
}($Command));
export { GetApplicationDateRangeKpiCommand };
