// @aws-sdk/client-pinpoint/dist-es/commands/GetInAppTemplateCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { GetInAppTemplateRequestFilterSensitiveLog, GetInAppTemplateResponseFilterSensitiveLog, } from "../models/models_1";
import { deserializeAws_restJson1GetInAppTemplateCommand, serializeAws_restJson1GetInAppTemplateCommand, } from "../protocols/Aws_restJson1";
var GetInAppTemplateCommand = (function (_super) {
    __extends(GetInAppTemplateCommand, _super);
    function GetInAppTemplateCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    GetInAppTemplateCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "PinpointClient";
        var commandName = "GetInAppTemplateCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: GetInAppTemplateRequestFilterSensitiveLog,
            outputFilterSensitiveLog: GetInAppTemplateResponseFilterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    GetInAppTemplateCommand.prototype.serialize = function (input, context) {
        return serializeAws_restJson1GetInAppTemplateCommand(input, context);
    };
    GetInAppTemplateCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_restJson1GetInAppTemplateCommand(output, context);
    };
    return GetInAppTemplateCommand;
}($Command));
export { GetInAppTemplateCommand };
