// @aws-sdk/client-pinpoint/dist-es/commands/GetInAppMessagesCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { GetInAppMessagesRequestFilterSensitiveLog, GetInAppMessagesResponseFilterSensitiveLog, } from "../models/models_1";
import { deserializeAws_restJson1GetInAppMessagesCommand, serializeAws_restJson1GetInAppMessagesCommand, } from "../protocols/Aws_restJson1";
var GetInAppMessagesCommand = (function (_super) {
    __extends(GetInAppMessagesCommand, _super);
    function GetInAppMessagesCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    GetInAppMessagesCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "PinpointClient";
        var commandName = "GetInAppMessagesCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: GetInAppMessagesRequestFilterSensitiveLog,
            outputFilterSensitiveLog: GetInAppMessagesResponseFilterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    GetInAppMessagesCommand.prototype.serialize = function (input, context) {
        return serializeAws_restJson1GetInAppMessagesCommand(input, context);
    };
    GetInAppMessagesCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_restJson1GetInAppMessagesCommand(output, context);
    };
    return GetInAppMessagesCommand;
}($Command));
export { GetInAppMessagesCommand };
