// @aws-sdk/client-pinpoint/dist-es/commands/PhoneNumberValidateCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { PhoneNumberValidateRequestFilterSensitiveLog, PhoneNumberValidateResponseFilterSensitiveLog, } from "../models/models_1";
import { deserializeAws_restJson1PhoneNumberValidateCommand, serializeAws_restJson1PhoneNumberValidateCommand, } from "../protocols/Aws_restJson1";
var PhoneNumberValidateCommand = (function (_super) {
    __extends(PhoneNumberValidateCommand, _super);
    function PhoneNumberValidateCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    PhoneNumberValidateCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "PinpointClient";
        var commandName = "PhoneNumberValidateCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: PhoneNumberValidateRequestFilterSensitiveLog,
            outputFilterSensitiveLog: PhoneNumberValidateResponseFilterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    PhoneNumberValidateCommand.prototype.serialize = function (input, context) {
        return serializeAws_restJson1PhoneNumberValidateCommand(input, context);
    };
    PhoneNumberValidateCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_restJson1PhoneNumberValidateCommand(output, context);
    };
    return PhoneNumberValidateCommand;
}($Command));
export { PhoneNumberValidateCommand };
