// @aws-sdk/client-pinpoint/dist-es/commands/UpdateApnsChannelCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { UpdateApnsChannelRequestFilterSensitiveLog, UpdateApnsChannelResponseFilterSensitiveLog, } from "../models/models_1";
import { deserializeAws_restJson1UpdateApnsChannelCommand, serializeAws_restJson1UpdateApnsChannelCommand, } from "../protocols/Aws_restJson1";
var UpdateApnsChannelCommand = (function (_super) {
    __extends(UpdateApnsChannelCommand, _super);
    function UpdateApnsChannelCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    UpdateApnsChannelCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "PinpointClient";
        var commandName = "UpdateApnsChannelCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: UpdateApnsChannelRequestFilterSensitiveLog,
            outputFilterSensitiveLog: UpdateApnsChannelResponseFilterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    UpdateApnsChannelCommand.prototype.serialize = function (input, context) {
        return serializeAws_restJson1UpdateApnsChannelCommand(input, context);
    };
    UpdateApnsChannelCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_restJson1UpdateApnsChannelCommand(output, context);
    };
    return UpdateApnsChannelCommand;
}($Command));
export { UpdateApnsChannelCommand };
