// @aws-sdk/client-pinpoint/dist-es/commands/UpdateSmsTemplateCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { UpdateSmsTemplateRequestFilterSensitiveLog, UpdateSmsTemplateResponseFilterSensitiveLog, } from "../models/models_1";
import { deserializeAws_restJson1UpdateSmsTemplateCommand, serializeAws_restJson1UpdateSmsTemplateCommand, } from "../protocols/Aws_restJson1";
var UpdateSmsTemplateCommand = (function (_super) {
    __extends(UpdateSmsTemplateCommand, _super);
    function UpdateSmsTemplateCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    UpdateSmsTemplateCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "PinpointClient";
        var commandName = "UpdateSmsTemplateCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: UpdateSmsTemplateRequestFilterSensitiveLog,
            outputFilterSensitiveLog: UpdateSmsTemplateResponseFilterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    UpdateSmsTemplateCommand.prototype.serialize = function (input, context) {
        return serializeAws_restJson1UpdateSmsTemplateCommand(input, context);
    };
    UpdateSmsTemplateCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_restJson1UpdateSmsTemplateCommand(output, context);
    };
    return UpdateSmsTemplateCommand;
}($Command));
export { UpdateSmsTemplateCommand };
