// @aws-sdk/client-pinpoint/dist-es/commands/UpdateTemplateActiveVersionCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { UpdateTemplateActiveVersionRequestFilterSensitiveLog, UpdateTemplateActiveVersionResponseFilterSensitiveLog, } from "../models/models_1";
import { deserializeAws_restJson1UpdateTemplateActiveVersionCommand, serializeAws_restJson1UpdateTemplateActiveVersionCommand, } from "../protocols/Aws_restJson1";
var UpdateTemplateActiveVersionCommand = (function (_super) {
    __extends(UpdateTemplateActiveVersionCommand, _super);
    function UpdateTemplateActiveVersionCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    UpdateTemplateActiveVersionCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "PinpointClient";
        var commandName = "UpdateTemplateActiveVersionCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: UpdateTemplateActiveVersionRequestFilterSensitiveLog,
            outputFilterSensitiveLog: UpdateTemplateActiveVersionResponseFilterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    UpdateTemplateActiveVersionCommand.prototype.serialize = function (input, context) {
        return serializeAws_restJson1UpdateTemplateActiveVersionCommand(input, context);
    };
    UpdateTemplateActiveVersionCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_restJson1UpdateTemplateActiveVersionCommand(output, context);
    };
    return UpdateTemplateActiveVersionCommand;
}($Command));
export { UpdateTemplateActiveVersionCommand };
