// @aws-sdk/client-pinpoint/dist-es/commands/UpdateInAppTemplateCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { UpdateInAppTemplateRequestFilterSensitiveLog, UpdateInAppTemplateResponseFilterSensitiveLog, } from "../models/models_1";
import { deserializeAws_restJson1UpdateInAppTemplateCommand, serializeAws_restJson1UpdateInAppTemplateCommand, } from "../protocols/Aws_restJson1";
var UpdateInAppTemplateCommand = (function (_super) {
    __extends(UpdateInAppTemplateCommand, _super);
    function UpdateInAppTemplateCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    UpdateInAppTemplateCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "PinpointClient";
        var commandName = "UpdateInAppTemplateCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: UpdateInAppTemplateRequestFilterSensitiveLog,
            outputFilterSensitiveLog: UpdateInAppTemplateResponseFilterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    UpdateInAppTemplateCommand.prototype.serialize = function (input, context) {
        return serializeAws_restJson1UpdateInAppTemplateCommand(input, context);
    };
    UpdateInAppTemplateCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_restJson1UpdateInAppTemplateCommand(output, context);
    };
    return UpdateInAppTemplateCommand;
}($Command));
export { UpdateInAppTemplateCommand };
