// @aws-sdk/client-lex-runtime-v2/dist-es/commands/RecognizeTextCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { RecognizeTextRequestFilterSensitiveLog, RecognizeTextResponseFilterSensitiveLog, } from "../models/models_0";
import { deserializeAws_restJson1RecognizeTextCommand, serializeAws_restJson1RecognizeTextCommand, } from "../protocols/Aws_restJson1";
var RecognizeTextCommand = (function (_super) {
    __extends(RecognizeTextCommand, _super);
    function RecognizeTextCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    RecognizeTextCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "LexRuntimeV2Client";
        var commandName = "RecognizeTextCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: RecognizeTextRequestFilterSensitiveLog,
            outputFilterSensitiveLog: RecognizeTextResponseFilterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    RecognizeTextCommand.prototype.serialize = function (input, context) {
        return serializeAws_restJson1RecognizeTextCommand(input, context);
    };
    RecognizeTextCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_restJson1RecognizeTextCommand(output, context);
    };
    return RecognizeTextCommand;
}($Command));
export { RecognizeTextCommand };
